File: src/LongestCommonPrefix.java
// https://leetcode.com/problems/longest-common-prefix/

public class LongestCommonPrefix {
    public String longestCommonPrefix(String[] array) {
        if (array.length == 0) {
            return "";
        }

        String result = array[0];
        for (int index = 1 ; index < array.length ; index++) {
            result = longestCommonPrefix(result, array[index]);
        }
        return result;
    }

    private String longestCommonPrefix(String first, String second) {
        for (int index = 0 ; index < first.length() && index < second.length() ; index++) {
            if (first.charAt(index) != second.charAt(index)) {
                return first.substring(0, index);
            }
        }
        return first.length() < second.length() ? first : second;
    }
}

File: python/longest_common_prefix.py
from typing import List


class Solution:
    def longestCommonPrefix(self, array: List[str]) -> str:
        if len(array) == 0:
            return ''

        result = array[0]
        for index in range(1, len(array)):
            result = self.longestCommonPrefix2Strings(result, array[index])
        return result

    def longestCommonPrefix2Strings(self, first: str, second: str) -> str:
        for index in range(min(len(first), len(second))):
            if first[index] != second[index]:
                return first[0:index]

        return first if len(first) < len(second) else second
